packages/client/src/client.ts

```diff
@@ -0,0 +1,165 @@
+import { jsonOptions, PenumbraService } from '@penumbra-zone/protobuf';
+import { createPromiseClient, PromiseClient } from '@connectrpc/connect';
+
+import { assertPenumbra, assertProviderManifest, assertProviderRecord } from './assert.js';
+import { PenumbraProviderNotConnectedError } from './error.js';
+import { PenumbraStateEventDetail, PenumbraStateEvent, isPenumbraStateEvent } from './event.js';
+import { availableOrigin } from './get.js';
+import { PenumbraProvider } from './provider.js';
+import { PenumbraState } from './state.js';
+import {
+  ChannelTransportOptions,
+  createChannelTransport,
+} from '@penumbra-zone/transport-dom/create';
+
+export interface IPenumbraClient {
+  /**
+   * Asks users to approve the connection to a specific browser manifest URL.
+   * If `manifest` argument is not provided, tries to connect to the first injected provider.
+   * Returns the manifest url of the connected provider, or an error
+   */
+  readonly connect: (providerUrl?: string) => Promise<string>;
+
+  /**
+   * Tries to reconnect to the injected provider without asking for user approval.
+   */
+  readonly reconnect: (providerUrl?: string) => Promise<string>;
+
+  /** Reexports the `disconnect` function from injected provider */
+  readonly disconnect: () => Promise<void>;
+  /** Reexports the `isConnected` function from injected provider  */
+  readonly isConnected: () => boolean | undefined;
+  /** Reexports the `state` function from injected provider */
+  readonly getState: () => PenumbraState;
+  /** Provides a simplified callback interface to `PenumbraStateEvent`s. */
+  readonly onConnectionChange: (cb: (connection: PenumbraStateEventDetail) => void) => void;
+
+  /** Returns a new or re-used `PromiseClient<T>` for a specific `PenumbraService` */
+  readonly service: <T extends PenumbraService>(
+    service: T,
+    options: Omit<ChannelTransportOptions, 'getPort'>,
+  ) => PromiseClient<T>;
+}
+
+export class PenumbraClient implements IPenumbraClient {
+  private readonly callbacks = new Set<(detail: PenumbraStateEventDetail) => void>();
+
+  private origin?: string;
+  private provider?: PenumbraProvider;
+  private port?: MessagePort;
+  private serviceClients: Map<string, PromiseClient<PenumbraService>>;
+
+  constructor() {
+    this.serviceClients = new Map();
+  }
+
+  private assertPort() {
+    if (!this.port) {
+      throw new PenumbraProviderNotConnectedError(this.origin);
+    }
+    return this.port;
+  }
+
+  private assertProvider() {
+    if (!this.provider) {
+      throw new PenumbraProviderNotConnectedError(this.origin);
+    }
+    return this.provider;
+  }
```

Comment:
given discussion requesting that the client avoid throwing, these methods were removed in #1648

notably, a `PenumbraClient` instance is useless without a selected provider. #1648 moved the origin parameter into the constructor - is there a good reason to revert that?

Reply:
good point, removed these assertion methods in favor of non-throwing.

as for the question – i will continue pursuing the client without the necessary `origin` in the constructor. It is not useless – it allows the creation of client in the setup of the users' app and the future reuse, a good practice popularized by many libraries (listed in the ADR as the source of inspiration). This approach is proven to be effective in terms of DX